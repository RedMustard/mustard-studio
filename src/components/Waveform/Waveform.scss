
.waveform__canvas {
    height: 100%;
    max-height: 100px;
    width: 100%;
}
