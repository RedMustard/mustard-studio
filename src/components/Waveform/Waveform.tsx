import { h } from 'preact';
import { useRef, useEffect } from 'preact/hooks';
import { WAVEFORM_BACKGROUND_COLOR, WAVEFORM_LINE_COLOR, WAVEFORM_LINE_WIDTH } from '../../constants';


interface WaveformProps {
    analyserNode: AnalyserNode;
}

export const Waveform = ({ analyserNode }: WaveformProps) => {
    const canvasRef = useRef(null);
    const bufferLength = analyserNode.frequencyBinCount;
    const dataArray = new Uint8Array(bufferLength);

    // Source: https://developer.mozilla.org/en-US/docs/Web/API/Web_Audio_API/Visualizations_with_Web_Audio_API#creating_a_waveformoscilloscope
    const drawWaveform = () => {
        const canvasWidth = canvasRef.current.width;
        const canvasHeight = canvasRef.current.height;
        const canvasContext = canvasRef.current.getContext('2d');
        const sliceWidth = (canvasWidth * 5.0) / bufferLength;

        analyserNode.getByteTimeDomainData(dataArray);

        canvasContext.fillStyle = WAVEFORM_BACKGROUND_COLOR;
        canvasContext.strokeStyle = WAVEFORM_LINE_COLOR;
        canvasContext.lineWidth = WAVEFORM_LINE_WIDTH;
        canvasContext.clearRect(0, 0, canvasWidth, canvasHeight);
        canvasContext.fillRect(0, 0, canvasWidth, canvasHeight);
        canvasContext.lineTo(canvasWidth, canvasHeight / 2);
        canvasContext.stroke();
        canvasContext.beginPath();

        let xCoord = 0;

        for (let i = 0; i < bufferLength; i++) {
            const v = dataArray[i] / 128.0;
            const yCoord = (v * canvasHeight) / 2;

            if (i === 0) {
                canvasContext.moveTo(xCoord, yCoord);
            } else {
                canvasContext.lineTo(xCoord, yCoord);
            }
            xCoord += sliceWidth;
        }

        requestAnimationFrame(drawWaveform);
    };

    useEffect(() => {
        drawWaveform();
    }, []);

    return (
        <div className="waveform">
            <canvas
                className="waveform__canvas"
                ref={canvasRef}
            />
        </div>
    );
};
