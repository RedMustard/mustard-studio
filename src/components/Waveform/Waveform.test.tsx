import { h } from 'preact';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Waveform } from './Waveform';

// eslint-disable-next-line import/newline-after-import
const wamock = require('web-audio-mock-api');
const audioContext = new wamock.AudioContext();

const baseProps = {
    analyserNode: audioContext.createAnalyser(),
};

describe('<Waveform />', () => {
    it('renders with basic props', () => {
        const wrapper = shallow(
            <Waveform
                {...baseProps}
            />,
        );
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
