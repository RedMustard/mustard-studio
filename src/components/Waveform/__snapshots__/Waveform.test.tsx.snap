// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Waveform /> renders with basic props 1`] = `
<div
  className="waveform"
>
  <canvas
    className="waveform__canvas"
  />
</div>
`;
